src/WebRTCSession.cpp

```diff
@@ -103,37 +79,154 @@ WebRTCSession::init(std::string *errorMessage)
 
 #ifdef GSTREAMER_AVAILABLE
 namespace {
-bool isoffering_;
+
+struct AudioSource
+{
+        std::string name;
+        GstDevice *device;
+};
+
+struct VideoSource
+{
+        struct Caps
+        {
+                std::string resolution;
+                std::vector<std::string> frameRates;
+        };
+        std::string name;
+        GstDevice *device;
+        std::vector<Caps> caps;
+};
+
 std::string localsdp_;
 std::vector<mtx::events::msg::CallCandidates::Candidate> localcandidates_;
-std::vector<std::pair<std::string, GstDevice *>> audioSources_;
+bool haveAudioStream_;
+bool haveVideoStream_;
+std::vector<AudioSource> audioSources_;
+std::vector<VideoSource> videoSources_;
+
+using FrameRate = std::pair<int, int>;
+std::optional<FrameRate>
+getFrameRate(const GValue *value)
+{
+        if (GST_VALUE_HOLDS_FRACTION(value)) {
+                gint num = gst_value_get_fraction_numerator(value);
+                gint den = gst_value_get_fraction_denominator(value);
+                return FrameRate{num, den};
+        }
+        return std::nullopt;
+}
+
+void
+addFrameRate(std::vector<std::string> &rates, const FrameRate &rate)
+{
+        constexpr double minimumFrameRate = 15.0;
+        if (static_cast<double>(rate.first) / rate.second >= minimumFrameRate)
+                rates.push_back(std::to_string(rate.first) + "/" + std::to_string(rate.second));
+}
+
+std::pair<int, int>
+tokenise(std::string_view str, char delim)
+{
+        std::pair<int, int> ret;
+        auto pos = str.find_first_of(delim);
+        auto s   = str.data();
+        std::from_chars(s, s + pos, ret.first);
+        std::from_chars(s + pos + 1, s + str.size(), ret.second);
```

Comment:
This breaks on gcc-7. I'm fine with dropping support for 7 and requiring gcc-8, but I'll let @redsky17 give an opinion on that too. Otherwise you may need to convert to a different string parsing function here.

Reply:
I guess the depends on whether we want to move our target ubuntu version for the AppImage build.  At least with official packages, GCC 8 isn't available until bionic (18.04).  Thoughts?

Reply:
Let's not sweat this. I'll parse this differently.